lib/tests/test_decorators.py

```diff
@@ -0,0 +1,114 @@
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from unittest import TestCase
+
+from muchos.config import decorators
+
+
+class DecoratedThing(object):
+    @property
+    @decorators.ansible_host_var
+    def host_var1(self):
+        return 'host_var1'
+
+    @property
+    @decorators.ansible_host_var(name='named_host_var2')
+    def host_var2(self):
+        return 'named_host_var2'
+
+    @property
+    @decorators.ansible_play_var
+    def play_var1(self):
+        return 'play_var1'
+
+    @property
+    @decorators.ansible_play_var(name='named_play_var2')
+    def play_var2(self):
+        return 'named_play_var2'
+
+    @property
+    @decorators.ansible_extra_var
+    def extra_var1(self):
+        return 'extra_var1'
+
+    @property
+    @decorators.ansible_extra_var(name='named_extra_var2')
+    def extra_var2(self):
+        return 'named_extra_var2'
+
+    @property
+    @decorators.default('default_val')
+    def default_val(self):
+        return None
+
+    @property
+    @decorators.required
+    def required_val(self):
+        return 'required_val'
+
+    @property
+    @decorators.required
+    def missing_required_val(self):
+        return None
+
+
+class DecoratorTests(TestCase):
+    def _flatten_dict(d):
+        return {(k, v) for k, v in d.items()}
+
+    def test_decorators(self):
+        thing = DecoratedThing()
+
+        all_host_vars = decorators.get_ansible_vars('host')
+        all_play_vars = decorators.get_ansible_vars('play')
+        all_extra_vars = decorators.get_ansible_vars('extra')
+
+        expected_host_vars = [
+            decorators._ansible_var('host_var1', 'DecoratedThing', 'host_var1'),
+            decorators._ansible_var('named_host_var2', 'DecoratedThing', 'host_var2')
+        ]
+
+        expected_play_vars = [
+            decorators._ansible_var('play_var1', 'DecoratedThing', 'play_var1'),
+            decorators._ansible_var('named_play_var2', 'DecoratedThing', 'play_var2')
+        ]
+
+        expected_extra_vars = [
+            decorators._ansible_var('extra_var1', 'DecoratedThing', 'extra_var1'),
+            decorators._ansible_var('named_extra_var2', 'DecoratedThing', 'extra_var2')
+        ]
+
+        self.assertEquals(
+            set([str(v) for v in expected_host_vars]) - set([str(v) for v in all_host_vars]),
```

Comment:
I am assuming this is doing set subtraction, but not completely sure.  Would it be better to check for a subset?   

I am thinking if `expected_host_vars` contained `[AV1,AV2]` and `all_host_vars` contained `[AV1,AV2,AV3]`,  then `expected_host_vars - all_host_vars` would be the empty set.  However it seems like it should fail because `all_host_vars` is not a subset of `expected_host_vars`.

Reply:
You are correct it is doing a set subtraction, however, this test is just validating that the things that were added are present, not that they're the only things present, e.g. expected_host_vars should be a subset of all_host_vars rather than the reverse.  It could be argued that given we have complete control that expected_host_vars should equal all_host_vars, but all_host_vars being a subset of expected_host_vars shouldn't be the check.

Reply:
it might be clearer if all_* was renamed actual_*, so actual_* is the test set, while expected_* is the validation/control set.

Reply:
The case I was thinking of was a really buggy change like the following to decorators.py.   This change add stuff where its wanted and also where its not wanted.  Would any of the unit test fail for this change?

```python
def ansible_var_decorator(var_type, name):
    def _decorator(func):
        ansible_var = _ansible_var(
            var_name=name if isinstance(name, str) else func.__name__,
            class_name=func.__qualname__.split('.')[0],
            property_name=func.__name__)
        _ansible_vars['host'].append(ansible_var)
        _ansible_vars['play'].append(ansible_var)
        _ansible_vars['extra'].append(ansible_var)
        return func

    if callable(name):
        return _decorator(name)
    return _decorator
```

Reply:
I'm picking this PR up on behalf of @sputnik13 - to respond to the last question, I checked and the unit tests as they stand right now will not pickup the case wherein the same variable was added to multiple variable groups. I can think through if / how we can catch that kind of fundamental issue with the code.

Reply:
This has now been addressed in the latest changes. Kindly review and hopefully this is good to go.

Reply:
It looks good @arvindshmicrosoft.  I was about to merge, but then I was not sure how you wanted to handle the two commits since there are multiple authors.  I would do a merge commit and not squash.

Reply:
OK sure, please go ahead and do a merge commit.